20250314-RS-event-streaming.md

```diff
@@ -0,0 +1,695 @@
+  oneof type {
+    WorkflowsEvent workflow = 1;      // From workflows.dapr.io/v1
+    ActorsEvent actor = 2;            // From actors.dapr.io/v1
+    JobsEvent job = 3;                // From jobs.dapr.io/v1
+    // add more below
+  }
+}
+```
+
+Each category (WorkflowsEvent, ActorsEvent, etc.) contains its own `oneof` with specific event types. For example:
+```protobuf
+message WorkflowsEvent {
+  oneof event_type {
+    // different workflow event types (see below)
+  }
+}
+```
+
+This nested `oneof` structure allows us to:
+- Group related events together (e.g., all workflow events)
+- Add new event types within each category
+- Maintain backward compatibility
+- Provide clear documentation of available events
+
+Please note that the event types within each category will grow as we add more events in Dapr.
+
+#### Future Considerations
+- For the scope of this proposal, events are currently organized by API domain:
+  - `workflows.dapr.io/v1` (e.g., WorkflowsEvent)
+  - `actors.dapr.io/v1` (e.g., ActorsEvent)
+  - `jobs.dapr.io/v1` (e.g., JobsEvent)
+  - etc
+- Future expansion to:
+  - Filter by specific event types within an API domain, example of how this could be done:
+    - `workflows.failed.dapr.io/v1`          // Only workflow failure events
+    - `workflows.completed.dapr.io/v1`       // Only workflow completion events
+    - `actors.activated.dapr.io/v1`          // Only actor activation events
+    - `components.initialized`               // Only component initialization events
+  - Component-related events (`components`) for monitoring component health and state
+    - Can start with (`components`) then further specify by component type (`components.<type>`)
+  - Control plane service events (would require control plane services to stream events to sidecars)
+    - Can add something like: `service.scheduler`, `service.placement`
+  - Health related events (`health`) for: dapr sidecar, control plane service, and component health related events
+  - Dapr related events (`dapr`) for: runtime config changes or application discovered events, etc 
+- A PubSub implementation could be added later if there's a strong need for guaranteed delivery, where users can configure a PubSub to be the event sink
+  - Where a new optional `type` field could be added where a user could put something like the following:
+    ```yaml
+    # PubSub sink type
+    - type: pubsub
+      component: "cassie's-pubsub"  # Reference to PubSub component
+      topic: "dapr-events"
+      metadata:  # Optional PubSub-specific configuration
+        partitionKey: "workflows.dapr.io/v1"  # Example: partition by event type
+      filters:
+        include: ["workflows.dapr.io/v1"]
+    ```
+- The current design allows for easy addition of new event types while maintaining type safety
+- Delivery options could be added to support both `best-effort` and `transactional` delivery by adding a `delivery` field, configurable per sink:
+  ```yaml
+  spec:
```

Comment:
Assuming [CloudEvent subscription spec](https://github.com/cloudevents/spec/blob/main/subscriptions/spec.md#321-subscription-object) this would be modified to put the delivery and security attributes under a "config" key and apply filters on a per-sink basis.

```yaml
events:
  subscriptions:
  - id: abc123
    sink: "audit-collector:8080"
    protocol: "http"
    config:
    -  delivery: "transactional" # or "best-effort" (default)
       security:
       - mtls: true
         identity: "spiffe://cluster.local/ns/cassie-app/sa/collector"
    filters: |
      exact:
      - type: io.dapr.workfows.failed
```